jax/_src/dtypes.py

```diff
@@ -416,8 +416,24 @@ def _type_promotion_lattice(jax_numpy_dtype_promotion: str) -> dict[JAXType, lis
   *f1_types, bf, f2, f4, f8 = _float_types
   c4, c8 = _complex_types
   i_, f_, c_ = _weak_types
-  if jax_numpy_dtype_promotion == 'standard':
-    out: dict[JAXType, list[JAXType]]
+  out: dict[JAXType, list[JAXType]]
+  if jax_numpy_dtype_promotion == 'full':
+    out = {
+      b1: [i_],
+      u1: [i2, u2], u2: [i4, u4], u4: [i8, u8], u8: [f_],
+      i_: [u1, i1], i1: [i2], i2: [i4], i4: [i8], i8: [f_],
+      f_: [*f1_types, bf, f2, c_],
+      **{t: [bf] for t in f1_types}, bf: [f4], f2: [f4], f4: [f8, c4], f8: [c8],
```

Comment:
Is this correct? Wouldn't f2 also be an acceptable promotion for all fp8 types we have?

Reply:
It would, but that would also break the lattice property. So if we want float8 to promote to both float16 and bfloat16, we'd lose associativity & would have to rewrite our dtype promotion infrastructure.

Reply:
Put another way, I'm curious to see the whole lattice and wondering if it is actually uniquely defined.

Reply:
Yes. So that implies to me that adding this particular promotion is broken, since it makes an arbitrary choice that isn't necessarily correct.

Reply:
> I'm curious to see the whole lattice and wondering if it is actually uniquely defined.

Wondering if what is uniquely defined?

Reply:
The lattice in this PR is essentially the same as pictured in https://jax.readthedocs.io/en/latest/type_promotion.html, except `f8` types lie on the connection between `f*` and `bf`.

Reply:
Yup, see my second comment. I don't think we can make that choice.

Reply:
Sorry, I don't follow

Reply:
Why is it ok to arbitrarily declare we will promote to `bfloat16` but not `float16`? It seems unprincipled.

Reply:
Also, use of a lattice itself is a design tradeoff. It has some nice properties, but if we get to the point where a promotion table makes more sense due to things like this, we could make that choice.

Reply:
Adding to this discussion. I would be a little disappointed if all fp8 types promoted to bf16. We (graphcore) have hardware in the wild that supports fp8, promotes to fp16, and doesn't natively support bf16.

Reply:
The only pragmatic bias I can think of would be to prefer promoting to IEEE types when possible. They will offer the broadest hardware support.

Reply:
Well, we have to choose to promote to either bf16 or f16, or otherwise abandon our lattice-based type promotion approach (why? well, if all fp8 flavors promote to both, then the promotion path for `fp8_flavor1 + fp8_flavor2` is undecidable on the lattice).

I chose bf16 here because it makes sense on hardware which has support for bf16 but not for f16.

Perhaps this is another argument in favor of requiring explicit promotion in mixed-precision ops with fp8?

Reply:
> But you have in mind other situations where f16 is the better choice.

The Graphcore IPU ([ISA here](https://docs.graphcore.ai/projects/isa/en/latest/index.html)) natively supports fp8, f16, and f32.
At 2.10.2.1 our ISA states:
> Note: Tile doesn’t provide direct support for the 16-bit BFloat16 truncated single format (8-bit exponent, 7-bit significand, Zero offset of 127). Use of this storage format is possible but requires explicit (zero tailed) conversion from/to full single-precision.

So my obvious bias would be towards promoting to f16, if all other factors are equal.
I'm not aware of any hardware that supports fp8, bf16, and not f16. Perhaps that will change.